### PR TITLE
FileChooser: "percent - unopened - finished last" consider status "complete" as 100%

### PR DESCRIPTION
**Pull request by hasezoey**
This PR adds onto the new filechooser sort `percent - unopened - finished last` to consider book status `complete` (Text: `Finished`) the same as 100%

re #11369

> `90% > 50% > 0% > unopened > 100% or finished`

https://github.com/koreader/koreader/pull/11369#issuecomment-1893897867

PS: it seems like the macos build is failing, but i do not see why

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11472)

### CODE REVIEW

**Review by Frenzie — APPROVED** [review]

**Comment by hius07** [comment]
Long ago opened books may do not have `percent_finished` or `summary` keys in doc_settings.
To avoid crash we do
https://github.com/koreader/koreader/blob/52fae11da79dda68caf634825f97272dc084b52d/frontend/ui/widget/filechooser.lua#L175
https://github.com/koreader/koreader/blob/52fae11da79dda68caf634825f97272dc084b52d/plugins/coverbrowser.koplugin/covermenu.lua#L59

**Review by Frenzie — CHANGES_REQUESTED** [review]
See comment, just changing so it's not accidentally merged

File: frontend/ui/widget/filechooser.lua
Comment (hasezoey) [inline frontend/ui/widget/filechooser.lua:240]: is this more correct? i am not really experienced in lua, does it maybe work similar to js's `&&` where `let t = a && b`  if `a` is truthy results in `t` getting the value of `b` (`t === b`) otherwise being `false` (or maybe in this case being `nil`)?

Comment (hasezoey) [reply frontend/ui/widget/filechooser.lua:240]: wait, i forgot i can simply test this, and yes, it is the same behavior as in js, so this could be a variable; should it?

Comment (Frenzie) [reply frontend/ui/widget/filechooser.lua:240]: The more obvious parallel with the conditional (`if (summary && summary.status) {}` comes down to the same thing.[1] I.e., if truthy, continue to the right, if also truthy, conditional as a whole true.

Assignment doesn't work the same: you'd have to write something like `local var = condition and true or false` if you wanted that specific outcome. Otherwise it's more akin to `let a = a ? a : b`.

[1] You could use `if (summary?.status)` these days, though I'm not quite convinced that optional chaining isn't an antipattern.

Comment (hius07) [reply frontend/ui/widget/filechooser.lua:240]: The condition is good.
You can move `percent_finished` (line 236) after this.

Comment (Frenzie) [reply frontend/ui/widget/filechooser.lua:240]: > wait, i forgot i can simply test this, and yes, it is the same behavior as in js, so this could be a variable; should it?

GH hadn't shown me this new reply while I was writing mine. I only just saw it via email. I would consider doing that undesirable, reducing clarity for no apparent reason. Clarity could potentially be improved that way when something else was also involved, but at least in this codebase we prefer to address that differently, for example along these lines:

```lua
if
  summary and summary.status
  and cover and cover.x and cover.y
  and hello and hello.goodbye
then
```

You could also write that as something like `summary_status and cover_x and cover_y and hello_goodbye` defined above but that's just not what we do here. ;-)

Though unlike using spaces instead of tabs, I also prefer it that way.

Comment (Frenzie) [reply frontend/ui/widget/filechooser.lua:240]: > The condition is good. You can move `percent_finished` (line 236) after this.

Grr, nor this reply. I thought GH updated the comments more or less live. Oh well.

Comment (hasezoey) [reply frontend/ui/widget/filechooser.lua:239]: i will mark this as resolved with "leave it as-is" and moving the "percent_finished" assignment after the early-return (done with c190668d6fe8b04d3773bcf6c7183f3a475ae196)

**Comment by Commodore64user** [comment]
one more comment to make after testing it, much like "finished" books, books marked as "on hold" appear to not respect any particular order. maybe it should be: `90% > 50% > 0% > **on hold** > unopened > 100% or finished`

**Comment by hasezoey** [comment]
> one more comment to make after testing it, much like "finished" books, books marked as "on hold" appear to not respect any particular order.

i knew about "on hold", but didnt know where it made the most sense, so i didnt add special handling (for now)

but i think `90% > 50% > 0% > on hold > unopened > 100% or finished` is reasonable, what do the maintainers think?

**Comment by Frenzie** [comment]
I'm not sure that special handling for on hold makes sense to me, it is effectively still in progress, but then again I use neither this sorting order nor on hold. For me on hold is a thing you might say about a book I haven't read for a few months but a special status for it sounds like a chore. So I guess perhaps its purpose would indeed be to get rid of it in a regular percentage listing?

**Comment by Commodore64user** [comment]
> but i think `90% > 50% > 0% > on hold > unopened > 100% or finished` is reasonable, what do the maintainers think?

so, is it going to work like this then? No one seems to have any suggestions …

**Pull request merged**
